### PR TITLE
Add POST task endpoint

### LINKED ISSUE #677: Add CREATE endpoint for tasks
Add POST endpoint for tasks.

Tasks have a few mandatory fields:

- date
- init
- end
- project 
- user

When creating a new task, we need to do a few checks:

- the end time needs to be after init.
- a task cannot overlap another existing task, that means a user shouldn't be allowed to create a tasks where the init or end times are between the init and end time of another task in the same day.
- user can create tasks in that date. This means the date is not blocked by the application admin. This is configured by the settings `dayLimitEnabled`, `numberOfDays` and `date`.
- project must be still active.

### PR DESCRIPTION
**Pull request by dmtrek14**
Closes #677 

- Update Pydantic to v2 and make associated syntax changes to several models
- Add validation, computed fields, and hybrid properties to support task creation
- Add services and service methods to validate tasks
- Add/update schemas for tasks

### CODE REVIEW

File: api/services/timelog.py [hunk 1]
```diff
@@ -48,3 +51,44 @@ def get_user_tasks(self, user_id: int, offset: int, limit: int, start: date, end
...
+        )
+        self.db.add(new_task)
+        self.db.commit()
+        self.db.refresh(new_task)
+        return new_task
+
+    def check_task_for_overlap(self, task: Task) -> TaskValidate:
+        validated_task = TaskValidate(is_task_valid=True, message="")
+        user_tasks_for_day = (
+            self.db.query(Task).where(Task.user_id == task.user_id, Task.date.between(task.date, task.date)).all() or []
```
Comment (anarute) [inline api/services/timelog.py:75]: Why between and not `Task.date == task.date`?

File: api/services/timelog.py [hunk 2]
```diff
@@ -48,3 +51,44 @@ def get_user_tasks(self, user_id: int, offset: int, limit: int, start: date, end
...
+                    validated_task.is_task_valid = False
+                    validated_task.message += f"You have already logged a task ending at {user_task_for_day.end_time}."
+                if task.end > user_task_for_day.init and task.init < user_task_for_day.end:
+                    validated_task.is_task_valid = False
+                    validated_task.message += (
+                        f"Task from {task.start_time} to {task.end_time} overlaps an existing task from"
+                        f" {user_task_for_day.start_time} to {user_task_for_day.end_time}."
+                    )
+
+        return validated_task
```
Comment (anarute) [inline api/services/timelog.py:94]:
```suggestion
        if len(user_tasks_for_day) <= 0:
            return validate_task

        for user_task_for_day in user_tasks_for_day:
            if task.init == user_task_for_day.init:
                validated_task.is_task_valid = False
                validated_task.message += f"You have already logged a task beginning at {user_task_for_day.start_time}."
            if task.end == user_task_for_day.end:
                validated_task.is_task_valid = False
                validated_task.message += f"You have already logged a task ending at {user_task_for_day.end_time}."
            if task.end > user_task_for_day.init and task.init < user_task_for_day.end:
                validated_task.is_task_valid = False
                validated_task.message += (
                    f"Task from {task.start_time} to {task.end_time} overlaps an existing task from"
                    f" {user_task_for_day.start_time} to {user_task_for_day.end_time}."
                )

        return validated_task
```

**Review by anarute — COMMENTED** [review]
Most of my comments are suggestions to start using early returns to decrease nesting and redundant checks.

File: api/routers/v1/timelog.py [hunk 3]
```diff
@@ -79,3 +84,56 @@ async def get_user_tasks(
...
+        validated.message += overlapping_tasks.message
+    if (
+        task_type_valid is True
+        and project_active is True
+        and user_can_create_tasks is True
+        and overlapping_tasks.is_task_valid is True
+    ):
+        validated.is_task_valid = True
+
+    return validated
```
[the remark this answers is not in the document]
Comment (dmtrek14) [reply api/routers/v1/timelog.py:139]: Makes good sense to me. Will update as you've suggested.

File: api/schemas/project.py [hunk 4]
```diff
@@ -1,22 +1,20 @@
...
-    project_type: Optional[str]
-    schedule_type: Optional[str]
+    init: Optional[date] = None
+    end: Optional[date] = None
+    invoice: Optional[float] = None
+    estimated_hours: Optional[float] = None
+    moved_hours: Optional[float] = None
+    description: Optional[str] = None
+    project_type: Optional[str] = None
+    schedule_type: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (dmtrek14) [reply api/schemas/project.py:17]: I think I can safely remove this. Pydantic has a helper tool to migrate from v1 to v2 and it made these changes. All the docs have the None as default, as well, so I probably just overlooked when committing. I will test and make sure we can remove.

File: api/schemas/timelog.py [hunk 5]
```diff
@@ -1,70 +1,130 @@
...
+    project_id: int
+    user_id: int
+    start_time: Annotated[str, StringConstraints(pattern=r"^[0-9]{2}:[0-9]{2}")]
+    end_time: Annotated[str, StringConstraints(pattern=r"^[0-9]{2}:[0-9]{2}")]
+
+    @computed_field
+    @property
+    def init(self) -> int:
+        hours, minutes = self.start_time.split(":")
+        return int(hours) * 60 + int(minutes)
```
[the remark this answers is not in the document]
Comment (dmtrek14) [reply api/schemas/timelog.py:105]: Good idea. Where should I put it? In a `helpers` folder inside `api`?

File: api/services/config.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            if not date_limit_enabled and not day_limit_enabled:
+                return True
+
+            if day_limit_enabled and number_of_days_limit is not None and number_of_days_limit > 0:
+                day_limit_date = date.today() + timedelta(days=number_of_days_limit)
+
+            if date_limit_enabled and cutoff_date is not None:
+                date_limit_date = cutoff_date
+
+            return task_date > day_limit_date or task_date > date_limit_date
```
[the remark this answers is not in the document]
Comment (dmtrek14) [reply api/services/config.py:27]: I like it and am happy to make the change.

File: api/services/timelog.py (hunk 1 above)
Comment (dmtrek14) [reply api/services/timelog.py:75]: No good reason. I think I had probably been copying the query from the `get_user_tasks` and didn't adjust this part.

File: api/schemas/project.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmtrek14) [reply api/schemas/project.py]: After some testing, I think it does need this even though it feels redundant. When I remove these, I start getting validation errors for story and description on create even though they still marked as optional.

File: api/schemas/timelog.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmtrek14) [reply api/schemas/timelog.py]: I went ahead and added a helper for time-related things and moved the two functions there that I used in the validation: `time_string_to_int` and `int_to_time_string`. I didn't have any smart ideas for the name of the file, so it is just named `time.py` and is in a new `api/helpers` folder. I am definitely open to better name ideas if you have one ;)

File: api/routers/v1/timelog.py [hunk 7]
```diff
@@ -79,3 +84,56 @@ async def get_user_tasks(
...
+    if current_user.id != task.user_id:
+        raise HTTPException(status_code=403, detail="You are not authorized to create tasks for this user.")
+    validated_task = validate_task(task, db)
+    if not validated_task.is_task_valid:
+        raise HTTPException(status_code=422, detail=validated_task.message)
+    result = TaskService(db).create_task(task)
+    return result
+
+
+def validate_task(task: TaskSchema, db: Session):
```
[the remark this answers is not in the document]
Comment (anarute) [reply api/routers/v1/timelog.py:115]: Right, ok let's leave it here then and if we find a better place for this kind of validation we move it :)

File: api/schemas/timelog.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (anarute) [reply api/schemas/timelog.py]: Sounds great, thank you!

**Review by anarute — APPROVED** [review]
It looks good to me, thank you for the changes!

**Pull request merged**